R/combineResults.R

```diff
@@ -0,0 +1,232 @@
+#'
+#' pred3 <- SingleR(test, ref1, labels=ref1$label, 
+#'     method="cluster", clusters=test$label) 
+#' pred4 <- SingleR(test, ref2, labels=ref2$label, 
+#'     method="cluster", clusters=test$label) 
+#'
+#' ###############################
+#' ##     Combining results     ##
+#' ###############################
+#'
+#' pred.single <- combineResults(list(pred1, pred2))
+#' pred.clust <- combineResults(list(pred3, pred4))
+#'
+#' @seealso
+#' \code{\link{matchReferences}}, to harmonize labels between reference datasets.
+#' \code{\link{SingleR}}, for generating predictions.
+#'
+#' @importFrom S4Vectors DataFrame metadata metadata<-
+#'
+#' @export
+combineResults <- function(results) {
+    num.features <- sapply(results, nrow)
+    if (abs(max(num.features) - min(num.features)) != 0) {
+        stop("Results DataFrames contain different numbers of cells or clusters.")
+    }
+
+    pred.scores <- list()
+    pred.firstlabels <- list()
+    pred.prunedlabels <- list()
+    pred.tuningscores <- list()
+    feature.names <- list()
+    de.genes <- list()
+    common.genes <- list()
+
+    # Keep track of which scores belong to which results object.
+    pred.score.indices <- list()
+    ind <- 1
+
+    for (i in seq_along(results)) {
+        res <- results[[i]]
+        pred.scores[[i]] <- res$scores
+
+        if (!is.null(res$first.labels)) {
+            pred.firstlabels[[i]] <- res$first.labels
+        } else {
+            pred.firstlabels[[i]] <- NA_character_
+        }
+
+        if (!is.null(res$pruned.labels)) {
+            pred.prunedlabels[[i]] <- res$pruned.labels
+        } else {
+            pred.prunedlabels[[i]] <- NA_character_
+        }
+
+        if (!is.null(res$tuning.scores)) {
+            pred.tuningscores[[i]] <- res$tuning.scores
+        } else {
+            pred.tuningscores[[i]] <- NA_character_
+        }
+
+        feature.names[[i]] <- rownames(res)
+
+        if (!is.null(metadata(res)$de.genes)) {
+            de.genes[[i]] <- metadata(res)$de.genes
+        } else {
+            de.genes[[i]] <- NA_character_
+        }
+
+        common.genes[[i]] <- metadata(res)$common.genes
+        pred.score.indices[[i]] <- seq(ind, (ind + ncol(res$scores) - 1))
+        ind <- ind + ncol(res$scores)
+    }
+
+    all.scores <- do.call(cbind, pred.scores)
+    best.labels <- colnames(all.scores)[max.col(all.scores)]
+
+    # Determine which results DataFrame each score comes from.
+    pred.indices <- sapply(max.col(all.scores), .get_pred_indices, pred.score.indices = pred.score.indices)
+
+    output <- DataFrame(scores = I(all.scores), labels = best.labels)
+
+    if (length(feature.names) > 0) {
+        if (length(unique(feature.names)) == 1) {
+            rownames(output) <- feature.names[[1]]
+        } else {
+            warning("Results Dataframes row names do not match, skipping row name assignment.")
+        }
+    }
+
+    if (length(unique(common.genes)) == 1) {
```

Comment:
Now that I'm actually investigating more closely, I think the common.genes metadata is somewhat misleading, as it appears it's really only the genes used in DE that are stuck there. These genes never match up between different references exactly, should we just stick each in a list? Maybe I'm not fully understanding.

As an example:

```
library(SingleR)
library(scRNAseq)
library(scater)
hpca.se <- HumanPrimaryCellAtlasData()
bp.se <- BlueprintEncodeData()

hESCs <- LaMannoBrainData('human-es')
hESCs <- hESCs[,1:100]

# Restrict to common genes between reference datasets:
common.refs <- intersect(rownames(bp.se), rownames(hpca.se))
hpca.se <- hpca.se[common.refs,]
bp.se <- bp.se[common.refs,]

# Restrict to common genes between test and reference data:
common <- intersect(rownames(hESCs), rownames(hpca.se))
hpca.se <- hpca.se[common,]
bp.se <- bp.se[common,]
hESCs <- hESCs[common,]
hESCs <- logNormCounts(hESCs)

pred.hpca <- SingleR(test = hESCs, ref = hpca.se, labels = hpca.se$label.main)
pred.bp <- SingleR(test = hESCs, ref = bp.se, labels = bp.se$label.main)
```

`pred.hpca` and `pred.bp` have different `common.genes` despite containing the same genes.

Reply:
Courtesy of #60, the different references will now be forced to have the same `common.genes`, otherwise it's difficult to compare the scores if they are computed from different genes.

For testing purposes, I guess you could disable this check, but they really should be the same moving forward.

Also, go to sleep.

Reply:
This is a warning in the latest commit just for testing purposes, but should be changed to a `stop` after #60 is pulled in.